README.md
# Uploadcare custom image loader for Next.js
![Test Status][test-status-img]
[![NPM version][npm-img]][npm-link]

<a href="https://uploadcare.com/?utm_source=github&utm_campaign=nextjs-loader">
  <img align="right" width="56" height="56"
    src="https://ucarecdn.com/1cc871de-5d82-442d-b4d6-aa2e35966879/-/resize/112x112/-/quality/lightest/logocircle2x.png"
    alt="">
</a>

The package helps you use the Uploadcare's transformation and CDN services from a Next.js app smoothly and easily.

It provides the `uploadcareLoader` function, which you can use as [a custom loader for the Next's Image component](https://nextjs.org/docs/api-reference/next/image#loader), and the `UploadcareImage` component with the custom loader enabled by default.

* [Demo](#demo)
* [Dependencies](#dependencies)
* [Installation](#installation)
* [Configuration](#configuration)
* [Usage](#usage)
* [Props](#props)
* [Notes](#notes)
* [Known Issues](#known-issues)
* [Links](#links)

## Demo

Look at the demo [here][demo-link].

## Dependencies

The only dependency is Next.js >= 10.0.5.

## Installation

```shell_script
yarn add @uploadcare/nextjs-loader
```

## Configuration

Inform Next that you're going to use a custom image loader through `next.config.js`:

```js
// next.config.js
module.exports = {
  images: {
    loader: "custom"
  }
}
```

Add your public Uploadcare key to your `.env*` config file. You can copy it from Dashboard -> API Keys -> Public key.

```ini
#.env
NEXT_PUBLIC_UPLOADCARE_PUBLIC_KEY="YOUR_PUBLIC_KEY"
```

Alternatively, in case you're using a [custom proxy endpoint][docs-custom-proxy-endpoint], set the proxy domain.

```ini
#.env
NEXT_PUBLIC_UPLOADCARE_CUSTOM_PROXY_DOMAIN="proxy.example.com"
```

Then set the app base url (to get relative image urls processed in Production):

```ini
#.env
NEXT_PUBLIC_UPLOADCARE_APP_BASE_URL="https://example.com/"
```

That's it. You may now use `@uploadcare/nextjs-loader` in your app (see [Usage](#usage)).

---

Below are optional parameters which you may not need:

```ini
#.env
# A comma-separated list of transformation parameters. Default: format/auto, stretch/off, progressive/yes
NEXT_PUBLIC_UPLOADCARE_TRANSFORMATION_PARAMETERS="format/auto, stretch/off, progressive/yes"
# Uploadcare CDN domain. Default: ucarecdn.com
NEXT_PUBLIC_UPLOADCARE_CUSTOM_CDN_DOMAIN="cdn.example.com"
```

> Please note `NEXT_PUBLIC_UPLOADCARE_TRANSFORMATION_PARAMETERS` override corresponding default parameters and keep others in place.

## Usage

**Option 1**. Use the `UploadcareImage` component and leave us the rest ;)
```tsx
import UploadcareImage from '@uploadcare/nextjs-loader';

<UploadcareImage
  alt="A test image"
  src="https://your-domain/image.jpg"
  width="400"
  height="300"
  quality="80"
/>
```
The `UploadcareImage` component supports the same parameters as the Next `Image` component.

**Option 2**. Pass `uploadcareLoader` to the `Image` component:

```tsx
import Image from 'next/image';
import { uploadcareLoader } from '@uploadcare/nextjs-loader';

<Image
  alt="A test image"
  src="https://your-domain/image.jpg"
  width="400"
  height="300"
  quality="80"
  loader={uploadcareLoader}
/>
```

**Option 3 (Next.js v13+ only)**. Use the [`loaderFile` setting][loader-file] to enable Uploadcare image loader for all images by default.

1. Configure the `loaderFile` in your `next.config.js` like the following:

```js
module.exports = {
  images: {
    loader: 'custom',
    loaderFile: './node_modules/@uploadcare/nextjs-loader/build/loader.js',
  },
}
```

2. Use `Image` as usual, with Uploadcare loader enabled implicitly:

```tsx
import Image from 'next/image';

<Image
  alt="A test image"
  src="https://your-domain/image.jpg"
  width="400"
  height="300"
  quality="80"
/>
```

**Option 4**. Use the [next-image-loader](https://www.npmjs.com/package/next-image-loader) plugin to enable Uploadcare image loader for all images by default

In that case, you may not need the `loader: "custom"` setting in your `next.config.js`.

1. Install [next-image-loader](https://www.npmjs.com/package/next-image-loader) and enable it as described in its README.

2. Create `image-loader.config.js` in the project root (in the same directory as `next.config.js`)
and add this code to it:

```js
// image-loader.config.js
import { imageLoader } from 'next-image-loader/build/image-loader';
import { uploadcareLoader } from '@uploadcare/nextjs-loader';

imageLoader.loader = uploadcareLoader;
```

3. Use `Image` as usual, with Uploadcare loader enabled implicitly:

```tsx
import Image from 'next/image';

<Image
  alt="A test image"
  src="https://your-domain/image.jpg"
  width="400"
  height="300"
  quality="80"
/>
```

Please note that you can still use any other loader for a separate image like this:

```tsx
import Image from 'next/image';
import anotherLoader from '[another-loader-project-name]';

<Image
  alt="A test image"
  src="https://your-domain/image.jpg"
  width="400"
  height="300"
  quality="80"
  loader={anotherLoader}
/>
```

where `anotherLoader` will be used instead of the Uploadcare loader for this particular image.

## Props

### `placeholder` and `blurDataURL`

[Read Next.js Image Component docs](https://nextjs.org/docs/api-reference/next/image#placeholder) about `placeholder` and `blurDataURL`.

There are two possible use cases:

  1. [When `src` is a string](#when-src-is-a-string)
  2. [When `src` is a static import](#when-src-is-a-static-import)

#### When `src` is a string

If you pass `placeholder="blur"` to the `Image` or `UploadcareImage` component, the `blurDataURL` property will be used as the placeholder. In this case you must provide the `blurDataURL` property using our `getBlurDataURL` server-side helper.

Here is the `getBlurDataURL` interface:

```ts
function getBlurDataURL(
  src: string,
  width = 10,
  quality = 1
): Promise<string>
```

Usage example:

```tsx
import UploadcareImage, { getBlurDataURL } from '@uploadcare/nextjs-loader';

const BLUR_IMAGE_URL = "https://your-domain/image.jpg"

export const getStaticProps = async () => {
  const blurDataURL = await getBlurDataURL(BLUR_IMAGE_URL);

  return {
    props: { blurDataURL }
  };
};

export default ({ blurDataURL }) => {
  return (
    <UploadcareImage
      alt="Blurred image"
      src={BLUR_IMAGE_URL}
      width="400"
      height="300"
      placeholder="blur"
      blurDataURL={blurDataURL}
    />
  )
}
```

#### When `src` is a static import

You can use both `UploadcareImage` and `Image` component with custom loader. In this case Next.js will generate base64 encoded image for `blurDataURL` automatically during build time.

```tsx
import staticImage from 'image.jpg'

<UploadcareImage
  alt="A test image"
  src={staticImage}
  width="400"
  height="300"
  quality="80"
  placeholder="blur"
/>
```

## Notes

- If you have jpegs larger than 3000px and you want loader to resize them up to 5000px, you need to pass filename with `jpeg` or `jpg`  extension to the src url. See [Output image dimensions docs](https://uploadcare.com/docs/transformations/image/#dimensions) for more details. When no filename provided, we'll treat the image as non-jpeg and can resize it up to 3000px only.

- If you pass a local image url, the loader returns it AS IS if the app is run in the development mode or if the `NEXT_PUBLIC_UPLOADCARE_APP_BASE_URL` is not set.

- Next.js Image component allows numeric only `quality` value. Uploadcare CDN supports string-based quality only (see [our docs](https://uploadcare.com/docs/transformations/image/compression/#operation-quality)). So numeric quality values will be mapped onto string ones using the following intervals:

  - 1-38 - lightest
  - 39-70 - lighter
  - 71-80 - normal
  - 81-88 - better
  - 89-100 - best

## Known Issues

**Issue 1:** Console warning like this:
> Image with src "${src}" has a "loader" property that does not implement width. Please implement it or use the "unoptimized" property instead.
Read more: https://nextjs.org/docs/messages/next-image-missing-loader-width

Next checks whether the image url which loader generates has the exact value which user passed through the `width` property of the `Image` component. And because the Uploadcare loader doesn't process SVG and GIF images, it just returns the same `src` value without adding any transformation parameters to it (including `width`). That's why Next reports the console warning.

**Fix:** Ignore the warning for now.

## Links

- [Uploadcare image compression documentation][uploadcare-transformation-image-compression-docs]
- [Uploadcare File Uploader for React](https://github.com/uploadcare/react-widget)

[test-status-img]: https://github.com/uploadcare/nextjs-loader/actions/workflows/test-and-lint.yml/badge.svg
[npm-img]: https://img.shields.io/npm/v/@uploadcare/nextjs-loader.svg
[npm-link]: https://www.npmjs.com/package/@uploadcare/nextjs-loader
[demo-link]: https://uploadcare-nextjs-loader.netlify.app/
[uploadcare-transformation-image-compression-docs]: https://uploadcare.com/docs/transformations/image/compression/?utm_source=github&utm_campaign=nextjs-loader
[docs-custom-proxy-endpoint]: https://uploadcare.com/docs/delivery/proxy/#usage-endpoint
[last-next-12-release]: https://github.com/uploadcare/nextjs-loader/tree/v0.4.0
[loader-file]: https://nextjs.org/docs/api-reference/next/image#loader-configuration


CHANGELOG.md
# [1.0.0](https://github.com/uploadcare/nextjs-loader/compare/v0.4.0...v1.0.0) (2022-11-22)


### Features

* add a `getBlurDataURL` server-side helper to generate base64 blurred placeholder ([5404d56](https://github.com/uploadcare/nextjs-loader/commit/5404d564e211aee6a633dfb272de71e0c4c0ce85))
* add separate `loader.js` module entry to be used with Next.js v13 `loaderFile` setting ([a6577b7](https://github.com/uploadcare/nextjs-loader/commit/a6577b7cca4c5a7b8ddf26beb29ad9ddc9a261f6))


### BREAKING CHANGES

* previously implemented `blurDataURL` auto-generation won't work in Next.js v13. It was a bad way because the blurry image was requested from the server in runtime. A new way is to generate a blurry image base64 string at build time using `getBlurDataURL`. See [README](https://github.com/uploadcare/nextjs-loader#when-src-is-a-string).



# [0.4.0](https://github.com/uploadcare/nextjs-loader/compare/v0.3.3...v0.4.0) (2022-06-08)


### Features

* auto-generate `blurDataURL` property when `placeholder=blur` ([bda8cf1](https://github.com/uploadcare/nextjs-loader/commit/bda8cf1545b3b856064e608eea49b1b18861a3db))



## [0.3.3](https://github.com/uploadcare/nextjs-loader/compare/v0.3.2...v0.3.3) (2022-06-01)


### Bug Fixes

* accept src without filename ([#14](https://github.com/uploadcare/nextjs-loader/issues/14)) ([7f18177](https://github.com/uploadcare/nextjs-loader/commit/7f1817794d7e84ccd0a049aba883ae223c6d1d46))



## [0.3.2](https://github.com/uploadcare/nextjs-loader/compare/v0.3.1...v0.3.2) (2022-01-21)


### Bug Fixes

* force format for large jpegs ([#11](https://github.com/uploadcare/nextjs-loader/issues/11)) ([2e93d47](https://github.com/uploadcare/nextjs-loader/commit/2e93d473884f7d808766d45cc8e71a6331cced57))



## [0.3.1](https://github.com/uploadcare/nextjs-loader/compare/v0.3.0...v0.3.1) (2021-11-09)


### Bug Fixes

* Fixed processing of the local image in the production mode. ([f165d3a](https://github.com/uploadcare/nextjs-loader/commit/f165d3a4caa486138954e77279ccb0efd48c4ad9))



# [0.3.0](https://github.com/uploadcare/nextjs-loader/compare/v0.2.0...v0.3.0) (2021-11-04)


### Features

* Bumped Next.js to 12. ([9c8be4c](https://github.com/uploadcare/nextjs-loader/commit/9c8be4c1a7abef88fda25690950f5f4c232c7329))



# [0.2.0](https://github.com/uploadcare/nextjs-loader/compare/v0.1.0...v0.2.0) (2021-11-03)


### Features

* Added support for local images. Improved automated tests. Made the demo consume the latest development snapshot. ([e495754](https://github.com/uploadcare/nextjs-loader/commit/e49575442005c74596c9cb6de605371878d629b9))



# [0.1.0](https://github.com/uploadcare/nextjs-loader/compare/v0.0.0...v0.1.0) (2021-10-28)

* Initial release





src/utils/helpers.ts
import {
  MAX_OUTPUT_IMAGE_DIMENSION,
  MAX_OUTPUT_JPEG_IMAGE_DIMENSION
} from './constants';

/**
 * Merge transformation parameters from `a` and `b`, so that `b` parameters have higher priority.
 *
 * @param {string[]} a
 * @param {string[]} b
 * @returns {string[]}
 */
export function mergeParams(a: string[], b: string[]): string[] {
  const resultParams = [...a];

  for (let i = 0; i < b.length; i++) {
    const [userParam] = _parseUploadcareTransformationParam(b[i]);

    let hasBeenReplaced = false;
    for (let j = 0; j < resultParams.length; j++) {
      if (resultParams[j].startsWith(userParam)) {
        resultParams[j] = b[i];
        hasBeenReplaced = true;
        break;
      }
    }

    // If the param is new, just add it.
    if (!hasBeenReplaced) {
      resultParams.push(b[i]);
    }
  }

  return resultParams;
}

/**
 * Get requested image format from params, for example jpeg or auto.
 *
 * @param {string[]} userParams
 * @returns {string}
 */
export function getRequestedFormatFromParams(userParams: string[]): string {
  for (let i = 0; i < userParams.length; i++) {
    const [key, value] = _parseUploadcareTransformationParam(userParams[i]);

    if (key === 'format') {
      return value;
    }
  }

  return 'auto';
}

export function getExtension(filename: string): string {
  return filename.toLowerCase().split('?')[0].split('#')[0].split('.')[1];
}

export function getFilename(url: string): string {
  return url.substring(1 + url.lastIndexOf('/'));
}

export function trimTrailingSlash(url: string): string {
  return url.replace(/\/$/, '');
}

export function convertToUploadcareQualityString(
  requestedQuality?: number
): string {
  // If any particular quality has not been requested, we use the smart quality mode.
  // zero quality is treated as no quality, so we use smart.
  // see https://nextjs.org/docs/api-reference/next/image#quality
  if (!requestedQuality) {
    return 'smart';
  }

  // Uploadcare doesn't support integer-based quality modificators,
  // so we need to map them onto uploadcare's equivalents
  const names = ['lightest', 'lighter', 'normal', 'better', 'best'];
  const intervals = [0, 38, 70, 80, 87, 100];
  const nameIdx = intervals.findIndex((min, idx) => {
    const max = intervals[idx + 1];
    return min <= requestedQuality && requestedQuality <= max;
  });
  return names[nameIdx];
}

/**
 * Get max resize width.
 *
 * Output image dimension is limited to 3000px,
 * but it can be increased by explicitly setting /format/jpeg/ through API params.
 *
 * @param {number} requestedWidth
 * @param {boolean} isForJpeg
 * @returns {number}
 */
export function getMaxResizeWidth(
  requestedWidth: number,
  isForJpeg = false
): number {
  const maxDimension = isForJpeg
    ? MAX_OUTPUT_JPEG_IMAGE_DIMENSION
    : MAX_OUTPUT_IMAGE_DIMENSION;
  return Math.min(Math.max(requestedWidth, 0), maxDimension);
}

export function generateDefaultProxyEndpoint(publicKey: string | null): string {
  return `https://${publicKey}.ucr.io`;
}

export function generateCustomProxyEndpoint(customProxyDomain: string): string {
  return `https://${customProxyDomain}`;
}

function isProtocolRelativeUrl(url: string): boolean {
  return url.startsWith('//');
}

export function isRelativeUrl(url: string): boolean {
  return url.startsWith('/') && !isProtocolRelativeUrl(url);
}

export function isCdnUrl(url: string, cdnDomain: string): boolean {
  if (isRelativeUrl(url)) {
    return false;
  }
  return new URL(url).hostname === cdnDomain.trim();
}

export function isProduction(): boolean {
  return process.env.NODE_ENV === 'production';
}

export function parseUserParamsString(paramsString?: string): string[] {
  if (paramsString == null || paramsString === '') {
    return [];
  }

  const params = paramsString.split(',');

  return params.map((param) => param.trim());
}

export function isDotenvParamEmpty(param: string | null): boolean {
  return param == null || param.trim() === '';
}

export function isJpegExtension(extension: string): boolean {
  return ['jpg', 'jpeg'].includes(extension.toLowerCase());
}

function _parseUploadcareTransformationParam(param: string): string[] {
  return param.split('/');
}
